src/execution.rs

```diff
@@ -114,6 +122,115 @@ pub trait Report {
+        };
+        self.report(&completed_task);
+    }
+}
+
+pub struct StageReportSender {
+    sender: pipe::Sender,
+    // Since we define the stages to be linear, we just need to
+    // keep one timestamp to report a stage's duration.
+    started_at: SystemTime,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum StageStatus {
+    Success,
+    Failure(i32),
+    Skipped(String),
+}
+
+impl From<StageStatus> for Status {
+    fn from(s: StageStatus) -> Self {
+        match s {
+            StageStatus::Success => Status::Success,
+            StageStatus::Failure(code) => Status::Failure(code),
+            StageStatus::Skipped(reason) => Status::Skipped(reason),
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Serialize, Deserialize)]
+pub struct StageReport {
+    stage_name: String,
+    status: StageStatus,
+    duration: Duration,
+}
+
+impl StageReportSender {
+    pub fn report_stage_status<N: ToString>(&mut self, stage_name: N, status: StageStatus) {
+        let stage_name = stage_name.to_string();
+        let end = std::time::SystemTime::now();
+        let start = self.started_at;
+        self.started_at = end;
+
+        let payload = StageReport {
+            stage_name,
+            status,
+            duration: end.duration_since(start).unwrap_or(Duration::from_secs(0)),
+        };
+
+        serialize_and_write(&mut self.sender, &payload).expect("Couldn't send");
+    }
+}
+
+fn serialize_and_write<W: Write, A: Serialize>(w: &mut W, payload: &A) -> io::Result<usize> {
+    let payload = bincode::serialize(payload).unwrap();
+    let n = w.write(&(payload.len() as usize).to_be_bytes())?;
+    let m = w.write(&payload)?;
+    Ok(n + m)
+}
+
+struct StreamDecoder {
+    buf: Vec<u8>,
+    offset: usize,
+}
+
+impl StreamDecoder {
+    fn new() -> Self {
+        StreamDecoder {
+            buf: Vec::new(),
+            offset: 0,
+        }
+    }
+
+    fn append(&mut self, data: &[u8]) {
+        self.buf.extend_from_slice(data);
+    }
+    // Decode a message if there is enough data in the buffer.
+    fn try_decode(&mut self) -> Option<StageReport> {
+        let avail = self.buf.len() - self.offset;
+
+        if avail > size_of::<usize>() {
+            let payload_size = &self.buf[self.offset..self.offset + size_of::<usize>()];
+            let payload_size = usize::from_be_bytes(payload_size.try_into().unwrap());
+
+            if avail < size_of::<usize>() + payload_size {
+                return None;
+            }
+
+            let payload = &self.buf
+                [self.offset + size_of::<usize>()..self.offset + size_of::<usize>() + payload_size];
+            let res: StageReport =
+                bincode::deserialize(&payload).unwrap_or_else(|e| panic!(format!("{:?}", e)));
+            // Update the offset
+            self.offset = self.offset + size_of::<usize>() + payload_size;
+            Some(res)
+        } else {
+            None
+        }
```

Comment:
I'd check it upfront to avoid extra nesting, i.e.

```rust
if aval < sizeof::<usize>() {
    return None;
}
// happy case
```

You do the same thing on line 219.

Reply:
Not that simple; I might have enough to read the size of the payload, but not have received all the bytes of the payload itself.

Reply:
Note that on line 219 we check that `aval < size_of::<usize>() + payload_size`

Reply:
done